crates/fj-host/src/host.rs

```diff
@@ -1,31 +1,119 @@
-use crossbeam_channel::Receiver;
+use std::thread;
 
-use crate::{Error, Evaluator, Model, ModelEvent, Watcher};
+use crossbeam_channel::{unbounded, Receiver, Sender};
+use fj_operations::shape_processor::ShapeProcessor;
+use winit::event_loop::EventLoopProxy;
 
-/// A Fornjot model host
+use crate::{HostCommand, HostHandle, Model, ModelEvent, Watcher};
+
+// Use a zero-sized error type to silence `#[warn(clippy::result_large_err)]`.
+// The only error from `EventLoopProxy::send_event` is `EventLoopClosed<T>`,
+// so we don't need the actual value. We just need to know there was an error.
+/// Error type for sending on a channel with a closed event loop
+pub struct EventLoopClosed;
+
+/// A Fornjot model host that runs in the background
 pub struct Host {
-    evaluator: Evaluator,
-    _watcher: Watcher,
+    shape_processor: ShapeProcessor,
+    event_loop_proxy: EventLoopProxy<ModelEvent>,
+    command_tx: Sender<HostCommand>,
+    command_rx: Receiver<HostCommand>,
 }
 
 impl Host {
-    /// Create a new instance of `Host`
-    ///
-    /// This is only useful, if you want to continuously watch the model for
-    /// changes. If you don't, just keep using `Model`.
-    pub fn from_model(model: Model) -> Result<Self, Error> {
-        let watch_path = model.watch_path();
-        let evaluator = Evaluator::from_model(model);
-        let watcher = Watcher::watch_model(watch_path, &evaluator)?;
-
-        Ok(Self {
-            evaluator,
-            _watcher: watcher,
-        })
+    /// Create a new `Host` that will process models for an event loop.
+    pub fn new(
+        shape_processor: ShapeProcessor,
+        event_loop_proxy: EventLoopProxy<ModelEvent>,
+    ) -> Self {
+        let (command_tx, command_rx) = unbounded();
+        Self {
+            shape_processor,
+            event_loop_proxy,
+            command_tx,
+            command_rx,
+        }
+    }
+
+    /// Run a background thread to watch for updates and process a model.
+    pub fn spawn(mut self) -> HostHandle {
+        let command_tx = self.command_tx.clone();
+
+        let host_thread = thread::Builder::new()
+            .name("host".to_string())
+            .spawn(move || -> Result<(), EventLoopClosed> {
+                let mut model: Option<Model> = None;
+                let mut _watcher: Option<Watcher> = None;
+
+                while let Ok(command) = self.command_rx.recv() {
+                    match command {
+                        HostCommand::LoadModel(new_model) => {
+                            // There's a race condition if a prior watcher sends
+                            // `TriggerEvaluation` right before this new watcher
+                            // is created.
```

Comment:
This probably can't happen right now, as `LoadModel` is only sent, if there isn't already an existing model. Does this sound reasonable?

Reply:
I think you're right that it couldn't happen with the code as-is, so the comment was for the future if there's e.g. a "Load new model" button in the gui or something similar.  Even then it would be highly unlikely and not bad even if it did happen.  It would be:
1. Old watcher sends TriggerEvaluation
2. New watcher created and new model evaluated
3. Old TriggerEvaluation received and new model evaluated again

Do you think that's the case, and if so, is it worth leaving a comment about?

Reply:
That makes sense!

I think it's worth expanding the comment. Can't really be too explicit about these kinds of things. Only a question of time until someone sees this and spends an hour trying to understand it :smile: